fairscale/optim/oss.py

```diff
@@ -382,16 +387,93 @@ def _broadcast_state_dict(self) -> None:
             else:
                 global_rank = self.get_global_rank(self.group, rank)
                 # Discard this tensor/rank, broadcast necessary for syncing
-                logging.debug("Discarding broadcast from rank %s", global_rank)
                 broadcast_object(empty_buffer, src_rank=global_rank, group=self.group, dist_device=self._device)
 
     def _free_other_grads(self) -> None:
         """Free all the gradients only useful for the other ranks
         """
-        for i, partition in enumerate(self.partition_parameters()):
-            if i == self.rank:
+        for rank, partition in enumerate(self.partition_parameters()):
+            if rank == self.rank:
                 continue
 
             for p in partition:
                 for t in p["params"]:
                     t.grad = None
+
+    @staticmethod
+    def get_global_rank(group: Any, rank: int) -> int:
+        if group is dist.group.WORLD:
+            return rank
+        else:
+            global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+        return global_rank
+
+    @staticmethod
+    def _broadcast_params(
+        buffers: List[torch.Tensor], per_rank_params: List[List[Parameter]], group: Any, self_rank: int
+    ) -> None:
+        """Helper function to broadcast all the parameters from a given device
+        """
+        buffer_size = buffers[0].numel()
+        restore_require_grad = []
+        bucket_requests = []
+        requests = []
+
+        # Bucket and issue all the async calls
+        for (rank, params), buffer in zip(enumerate(per_rank_params), buffers):
+            # All the params are sorted per rank and per increasing size
+            if len(params) == 0:
+                continue
+
+            global_rank = OSS.get_global_rank(group, rank)
+
+            # Copy small parameters into per-GPU buffers
+            i_bucketed = 0  # the number of tensors packed in the buffer
+            offset = 0
+
+            # Since all the parameters are already sorted per increasing size, we only need to consider the first ones.
+            while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
+                end = offset + params[i_bucketed].numel()
+                if rank == self_rank:
```

Comment:
@blefaudeux  Naman and I were trying to figure out why we were getting different results on Fairscale when using model parallelism. It looks like this diff broke the behavior I had fixed. 

Proprer behavior is restored if you change this line to `if rank == global_rank:`

cc @ngoyal2707 who is opening a fix

Reply:
oh, sorry about that :( could we add a proper unit test or regression test so that I don't break things again ?

Reply:
The global_rank code is subtle, a unit test that prevent regression is non-trivial but probably very much worth it.

Reply:
@min-xu-ai I don't have a unit test, but I do have a regression/integration test. It's described in the PR description here: https://github.com/facebookresearch/fairscale/pull/121

Basically, run fairseq w/ and w/o OSS and check if the training logs are the same.

Reply:
Don't worry. I was thinking a unit test within this repo since we change the code here from time to time and we don't want to wait until your test run to find out regression.

Reply:
It could still be added here! Run the script and compare the output. Sounds like Benjamin has a unit test though